### PR TITLE
Adds support for FOTA for following targets 1. DISCO_L475VG_IOT01A 2. K64F 3. K66F 4. NUCLEO_F411RE 5. NUCLEO_F429ZI.

### PR DESCRIPTION
**Pull request by arpitarm**
### Summary of changes 

[] I confirm this contribution is my own and I agree to license it with Apache 2.0.

[] I confirm the moderators may change the PR before merging it in.

For new board enablements only:

[] I confirm the board is [Mbed Enabled](https://www.mbed.com/en/about-mbed/mbed-enabled/introduction/) and passes the Mbed Enabled test set.

[] I confirm the contribution has been tested properly and the tests results for [TESTS](https://github.com/ARMmbed/mbed-os-example-pelion/tree/master/TESTS) are attached.

[] I confirm `mbed-os.lib` and `mbed-cloud-client.lib` hashes or the content in folders `mbed-os` and `mbed-cloud-client` were not modified in order to pass the tests.

- Maintainers of this repository update the Mbed OS and Client hashes.
- Please report any issues through issues in relevant repositories.

### CODE REVIEW

**Comment by iot-pdmc** [comment]
Thanks for your changes.
@ARMmbed/pdmc-integrators please review.

File: mbed_cloud_client_user_config.h [hunk 1]
```diff
@@ -34,7 +34,7 @@
 #endif
 
 /* set flag to enable update support in mbed Cloud client */
-#define MBED_CLOUD_CLIENT_SUPPORT_UPDATE
+#undef MBED_CLOUD_CLIENT_SUPPORT_UPDATE
```
Comment (teetak01) [inline mbed_cloud_client_user_config.h:37]: Create a new mbed_cloud_client_user_config_legacy.h file for legacy targets.

File: mbed_app.json [hunk 2]
```diff
@@ -1,71 +1,68 @@
...
-        "MBED_CLIENT_USER_CONFIG_FILE=\"mbed_cloud_client_user_config.h\"",
         "MBED_CLOUD_CLIENT_USER_CONFIG_FILE=\"mbed_cloud_client_user_config.h\"",
         "ARM_UC_USE_PAL_BLOCKDEVICE=1",
         "PAL_PLATFORM_DEFINED_CONFIGURATION=\"mbedOS_SST.h\"",
         "PAL_USER_DEFINED_CONFIGURATION=\"mbedOS_SST.h\""
     ],
     "target_overrides": {
         "*": {
-            "target.features_add"                       : ["BOOTLOADER", "STORAGE"],
+            "target.features_add"                       : ["STORAGE"],
```
[the remark this answers is not in the document]
Comment (arpitarm) [reply mbed_app.json:11]: Done

File: mbed_app.json [hunk 3]
```diff
@@ -1,71 +1,68 @@
...
             "platform.stdio-baud-rate"                  : 115200,
             "platform.stdio-convert-newlines"           : true,
             "platform.stdio-buffered-serial"            : true,
             "platform.stdio-flush-at-exit"              : true,
             "rtos.main-thread-stack-size"               : 5120,
-            "update-client.storage-locations"           : 1,
-            "mbed-trace.enable"                         : null,
             "events.shared-stacksize"                   : 2048,
+            "events.shared-eventsize"                   : 2048,
+            "update-client.storage-locations"           : 1,
```
[the remark this answers is not in the document]
Comment (arpitarm) [reply mbed_app.json:20]: Done

File: mbed_app.json [hunk 4]
```diff
@@ -1,71 +1,68 @@
...
+          "target.macros_add"                        : ["MBEDTLS_USER_CONFIG_FILE=\"mbedTLSConfig_mbedOS.h\""],
+          "target.network-default-interface-type"     : "ETHERNET",
+          "target.extra_labels_add"                   : ["BL_INTERNAL_FLASH"],
+          "target.header_offset"                      : "0xa000",
+          "target.app_offset"                         : "0xa200",
+          "target.restrict_size"                      : "0xF7E00",
+          "target.components_remove"                  : ["SD"],
+          "fota.enable"                               : true,
+          "fota.default-app-ifs"                      : 1,
+          "fota.num-components"                       : 2,
```
[the remark this answers is not in the document]
Comment (arpitarm) [reply mbed_app.json:46]: Done

File: mbed_app.json [hunk 5]
```diff
@@ -1,71 +1,68 @@
...
-            "storage.storage_type"                      : "TDB_INTERNAL",
-            "run-ci"                                    : 2
+          "target.macros_add"                        : ["MBEDTLS_USER_CONFIG_FILE=\"mbedTLSConfig_mbedOS.h\""],
+          "target.network-default-interface-type"     : "ETHERNET",
+          "target.extra_labels_add"                   : ["BL_INTERNAL_FLASH"],
+          "target.header_offset"                      : "0x8000",
+          "target.app_offset"                         : "0x8200",
+          "target.restrict_size"                      : "0xF7E00",
+          "target.components_remove"                  : ["SD"],
+          "mbed-cloud-client.enable-device-sentry"    : true,
```
[the remark this answers is not in the document]
Comment (arpitarm) [reply mbed_app.json:57]: Done

File: mbed_app.json [hunk 6]
```diff
@@ -1,71 +1,68 @@
...
+          "nsapi.socket-stats-enabled"                : true,
+          "platform.cpu-stats-enabled"                : 1,
+          "platform.heap-stats-enabled"               : 1,
```
[the remark this answers is not in the document]
Comment (arpitarm) [reply mbed_app.json:60]: Done

File: mbed_cloud_client_user_config.h (hunk 1 above)
Comment (arpitarm) [reply mbed_cloud_client_user_config.h:37]: Done

File: mbed_app.json [hunk 7]
```diff
@@ -1,71 +1,68 @@
...
-            "update-client.storage-address"             : "(MBED_CONF_STORAGE_TDB_INTERNAL_INTERNAL_BASE_ADDRESS+MBED_CONF_STORAGE_TDB_INTERNAL_INTERNAL_SIZE)",
-            "update-client.storage-size"                : "(512*1024-MBED_CONF_STORAGE_TDB_INTERNAL_INTERNAL_SIZE)",
-            "update-client.storage-locations"           : 1,
-            "update-client.storage-page"                : 8,
-            "mbed-cloud-client.update-storage"          : "ARM_UCP_FLASHIAP",
-            "storage_tdb_internal.internal_base_address": "(512*1024)",
-            "storage_tdb_internal.internal_size"        : "(32*1024)",
-            "storage.storage_type"                      : "TDB_INTERNAL",
-            "run-ci"                                    : 2
+          "target.macros_add"                        : ["MBEDTLS_USER_CONFIG_FILE=\"mbedTLSConfig_mbedOS.h\""],
```
[the remark this answers is not in the document]
Comment (arpitarm) [reply mbed_app.json:37]: "client_app.user-config"                    : "\"mbed_cloud_client_user_config_legacy.h\"",

**Comment by anttiylitokola** [comment]
run-ci

**Comment by iot-pdmc** [comment]
# [Test run: FAILED](https://jenkins-client.isgtesting.com/job/mbed-os-example-pelion/748/)
***

**Test summary:**
Executed: 0
Failed: 0
Skipped: 0
Errors: 0
[Build artifacts](http://pdmc-ci.s3-website-eu-west-1.amazonaws.com/?prefix=jenkins-ci/mbed-os-example-pelion/748)

posted-by-tag: https://jenkins-client.isgtesting.com/job/mbed-os-example-pelion/

File: mbed_app.json [hunk 8]
```diff
@@ -113,9 +88,14 @@
...
             "esp8266.rts"                               : "PA_12",
             "esp8266.cts"                               : "PA_11",
-            "esp8266.provide-default"                   : true
+            "esp8266.provide-default"                   : true,
+            "fota.enable"                               : true,
+            "fota.default-app-ifs"                      : 1,
+            "fota.app-default-cb-no-print"              : false
         },
         "NUMAKER_PFM_NUC472": {
+            "client_app.user-config"                    : "\"mbed_cloud_client_user_config_legacy.h\"",
```
Comment (teetak01) [inline mbed_app.json:97]: Just add these to macros_add, instead of adding application mbed_lib.json

File: mbed_app.json [hunk 9]
```diff
@@ -1,109 +1,84 @@
...
-            "storage_tdb_internal.internal_base_address": "(MBED_ROM_START+1024*1024)",
-            "storage_tdb_internal.internal_size"        : "(128*1024)",
-            "storage.storage_type"                      : "TDB_INTERNAL"
+            "target.app_offset"                         : "0x8200",
+            "target.restrict_size"                      : "0xF7E00",
+            "fota.enable"                               : true,
+            "fota.default-app-ifs"                      : 1,
+            "fota.app-default-cb-no-print"              : false
         },
         "NUCLEO_F411RE": {
```
Comment (teetak01) [inline mbed_app.json:72]: STORAGE missing?

**Comment by arpitarm** [comment]
run-ci

**Comment by iot-pdmc** [comment]
# [Test run: FAILED](https://jenkins-client.isgtesting.com/job/mbed-os-example-pelion/749/)
***

**Test summary:**
Executed: 0
Failed: 0
Skipped: 0
Errors: 0
[Build artifacts](http://pdmc-ci.s3-website-eu-west-1.amazonaws.com/?prefix=jenkins-ci/mbed-os-example-pelion/749)

posted-by-tag: https://jenkins-client.isgtesting.com/job/mbed-os-example-pelion/

**Comment by arpitarm** [comment]
run-ci

**Comment by iot-pdmc** [comment]
# [Test run: FAILED](https://jenkins-client.isgtesting.com/job/mbed-os-example-pelion/750/)
***

**Test summary:**
Executed: 0
Failed: 0
Skipped: 0
Errors: 0

**Failed to build following targets:**
NUMAKER_IOT_M487-ARM-TCP
NUMAKER_PFM_NUC472-GCC_ARM-TCP
NUMAKER_IOT_M263A-GCC_ARM-TCP
NUMAKER_PFM_NUC472-ARM-TCP
DISCO_L496AG-ARM-TCP
DISCO_F746NG-GCC_ARM-TCP
NUMAKER_IOT_M487-GCC_ARM-TCP
NUMAKER_PFM_M487-GCC_ARM-TCP
EP_AGORA-ARM-TCP
NUMAKER_PFM_M487-ARM-TCP
NUMAKER_IOT_M263A-ARM-TCP
DISCO_L496AG-GCC_ARM-TCP
DISCO_F746NG-ARM-TCP
ARCH_MAX-ARM-TCP
EP_AGORA-GCC_ARM-TCP
ARCH_MAX-GCC_ARM-TCP
CY8CPROTO_062_4343W-ARM-TCP
CY8CPROTO_062_4343W-GCC_ARM-TCP
[Build artifacts](http://pdmc-ci.s3-website-eu-west-1.amazonaws.com/?prefix=jenkins-ci/mbed-os-example-pelion/750)

posted-by-tag: https://jenkins-client.isgtesting.com/job/mbed-os-example-pelion/

**Comment by Mayankbht** [comment]
run-ci

File: mbed_app.json [hunk 10]
```diff
@@ -411,7 +394,10 @@
             "target.macros_remove"                      : ["MBEDTLS_CONFIG_HW_SUPPORT"]
         },
         "RZ_A1H": {
-            "target.macros_add"                         : ["MBEDTLS_USER_CONFIG_FILE=\"mbedTLSConfig_mbedOS_SW_TRNG.h\"", "MBEDTLS_ENTROPY_NV_SEED", "PAL_USE_HW_TRNG=0"],
+            "events.shared-eventsize"                   : 2048,
```
Comment (teetak01) [inline mbed_app.json:397]: no need to repeat the events here. Its already on top-section.

File: mbed_app.json [hunk 11]
```diff
@@ -430,7 +416,10 @@
             "storage.storage_type"                      : "TDB_INTERNAL"
         },
         "GR_LYCHEE": {
-            "target.macros_add"                         : ["MBEDTLS_USER_CONFIG_FILE=\"mbedTLSConfig_mbedOS.h\""],
+            "events.shared-eventsize"                   : 2048,
```
Comment (teetak01) [inline mbed_app.json:419]: no need to repeat the events here. Its already on top-section.

File: mbed_app.json [hunk 12]
```diff
@@ -450,8 +439,10 @@
             "storage.storage_type"                      : "TDB_INTERNAL"
         },
         "WIO_3G": {
+            "events.shared-eventsize"                   : 2048,
```
Comment (teetak01) [inline mbed_app.json:442]: no need to repeat the events here. Its already on top-section.

File: mbed_app.json [hunk 13]
```diff
@@ -479,8 +470,10 @@
             "storage.storage_type"                      : "TDB_INTERNAL"
         },
         "WIO_BG96": {
+            "events.shared-eventsize"                   : 2048,
```
Comment (teetak01) [inline mbed_app.json:473]: no need to repeat the events here. Its already on top-section.

File: mbed_app.json [hunk 14]
```diff
@@ -507,8 +500,9 @@
             "storage.storage_type"                      : "TDB_INTERNAL"
         },
         "UHURU_RAVEN": {
+            "target.features_add"                       : ["BOOTLOADER", "STORAGE"],
```
Comment (teetak01) [inline mbed_app.json:503]: "update-client.storage-locations"           : 1,
            
            missing here?

File: mbed_app.json [hunk 15]
```diff
@@ -534,8 +528,9 @@
             "storage_tdb_internal.internal_size"        : "(32*2*1024)"
         },
         "NUCLEO_F767ZI": {
+            "target.features_add"                       : ["BOOTLOADER", "STORAGE"],
```
Comment (teetak01) [inline mbed_app.json:531]: "update-client.storage-locations"           : 1,

missing ?

File: mbed_app.json [hunk 16]
```diff
@@ -554,7 +549,10 @@
             "storage_tdb_internal.internal_size"        : "(32*2*1024)"
         },
         "CY8CPROTO_062_4343W": {
-            "target.macros_add"                         : ["MBEDTLS_USER_CONFIG_FILE=\"mbedTLSConfig_mbedOS.h\"","ARM_UC_DELTAPAAL_WRITE_BUF_SIZE=512"],
+            "events.shared-eventsize"                   : 2048,
```
Comment (teetak01) [inline mbed_app.json:552]: no need to repeat the events here

File: mbed_app.json [hunk 17]
```diff
@@ -590,6 +588,6 @@
             "help"           : "Configuration item to control the level of CI verification. 0 = ignored, 1 = build-only, 2 = build + E2E tests",
             "accepted_values": [0, 1, 2],
             "value"          : 1
-        }
+    }
```
Comment (teetak01) [inline mbed_app.json:591]: intendation

File: mbed_app.json [hunk 18]
```diff
@@ -1,109 +1,85 @@
 {
     "macros": [
         "MBED_TRACE_MAX_LEVEL=TRACE_LEVEL_INFO",
```
Comment (teetak01) [inline mbed_app.json:3]: This line can be removed now. Mbed OS6 has this enabled by default.

Comment (teetak01) [inline mbed_app.json:3]:
```suggestion
```
This line can be removed now. Mbed OS6 has this enabled by default.

**Comment by arpitarm** [comment]
run-ci

**Comment by iot-pdmc** [comment]
# [Test run: SUCCESS](https://jenkins-client.isgtesting.com/job/mbed-os-example-pelion/751/)
***

**Test summary:**
Executed: 0
Failed: 0
Skipped: 0
Errors: 0
[Build artifacts](http://pdmc-ci.s3-website-eu-west-1.amazonaws.com/?prefix=jenkins-ci/mbed-os-example-pelion/751)

posted-by-tag: https://jenkins-client.isgtesting.com/job/mbed-os-example-pelion/

**Comment by Mayankbht** [comment]
run-ci

**Comment by arpitarm** [comment]
run-ci

File: mbed_app.json (hunk 18 above)
Comment (Mayankbht) [reply mbed_app.json]: Done

Comment (Mayankbht) [reply mbed_app.json]: Done

File: mbed_app.json (hunk 17 above)
Comment (Mayankbht) [reply mbed_app.json]: Done

File: mbed_app.json (hunk 16 above)
Comment (Mayankbht) [reply mbed_app.json]: Done

File: mbed_app.json (hunk 15 above)
Comment (Mayankbht) [reply mbed_app.json:524]: Already in place.

File: mbed_app.json (hunk 14 above)
Comment (Mayankbht) [reply mbed_app.json:496]: Already in place.

File: mbed_app.json (hunk 13 above)
Comment (Mayankbht) [reply mbed_app.json]: Done.

File: mbed_app.json (hunk 12 above)
Comment (Mayankbht) [reply mbed_app.json]: Done.

File: mbed_app.json (hunk 11 above)
Comment (Mayankbht) [reply mbed_app.json]: Done.

File: mbed_app.json (hunk 10 above)
Comment (Mayankbht) [reply mbed_app.json]: Done.

File: mbed_app.json [hunk 19]
```diff
@@ -1,109 +1,85 @@
...
-            "update-client.storage-page"                : 1,
-            "mbed-cloud-client.update-storage"          : "ARM_UCP_FLASHIAP",
-            "storage_tdb_internal.internal_base_address": "(MBED_ROM_START+1024*1024)",
-            "storage_tdb_internal.internal_size"        : "(128*1024)",
-            "storage.storage_type"                      : "TDB_INTERNAL"
+            "target.app_offset"                         : "0x8200",
+            "target.restrict_size"                      : "0xF7E00",
+            "fota.enable"                               : true,
+            "fota.default-app-ifs"                      : 1,
+            "fota.app-default-cb-no-print"              : false
```
[the remark this answers is not in the document]
Comment (Mayankbht) [reply mbed_app.json]: Done.

File: mbed_app.json [hunk 20]
```diff
@@ -1,109 +1,85 @@
...
-            "storage_tdb_internal.internal_base_address": "(1024*1024)",
-            "storage_tdb_internal.internal_size"        : "(96*1024)",
-            "storage.storage_type"                      : "TDB_INTERNAL",
-            "run-ci"                                    : 2
+            "target.app_offset"                         : "0x8200",
+            "target.restrict_size"                      : "0xF7E00",
+            "target.components_remove"                  : ["SD"],
+            "fota.enable"                               : true,
+            "fota.default-app-ifs"                      : 1,
+            "fota.app-default-cb-no-print"              : false
```
[the remark this answers is not in the document]
Comment (Mayankbht) [reply mbed_app.json]: Done.

File: mbed_app.json [hunk 21]
```diff
@@ -1,109 +1,85 @@
...
-            "storage_tdb_internal.internal_base_address": "(512*1024)",
-            "storage_tdb_internal.internal_size"        : "(32*1024)",
-            "storage.storage_type"                      : "TDB_INTERNAL",
-            "run-ci"                                    : 2
+            "target.app_offset"                         : "0x8200",
+            "target.restrict_size"                      : "0xF7E00",
+            "target.components_remove"                  : ["SD"],
+            "fota.enable"                               : true,
+            "fota.default-app-ifs"                      : 1,
+            "fota.app-default-cb-no-print"              : false
```
[the remark this answers is not in the document]
Comment (Mayankbht) [reply mbed_app.json]: Done.

File: mbed_app.json [hunk 22]
```diff
@@ -1,109 +1,85 @@
...
+            "storage_filesystem.external_base_address"  : null,
+            "storage_filesystem.external_size"          : null,
             "nsapi.default-wifi-security"               : "WPA_WPA2",
             "nsapi.default-wifi-ssid"                   : "\"SSID\"",
-            "nsapi.default-wifi-password"               : "\"PASSWORD\"",
-            "nsapi.default-cellular-apn"                : "\"APN\"",
-            "nsapi.default-cellular-sim-pin"            : null,
-            "nsapi.default-cellular-username"           : null,
-            "nsapi.default-cellular-password"           : null
+            "nsapi.default-wifi-password"               : "\"Password\""
```
[the remark this answers is not in the document]
Comment (Mayankbht) [reply mbed_app.json]: Done.

File: mbed_app.json [hunk 23]
```diff
@@ -1,109 +1,85 @@
 {
     "macros": [
         "MBED_TRACE_MAX_LEVEL=TRACE_LEVEL_INFO",
-        "MBED_CLIENT_USER_CONFIG_FILE=\"mbed_cloud_client_user_config.h\"",
         "MBED_CLOUD_CLIENT_USER_CONFIG_FILE=\"mbed_cloud_client_user_config.h\"",
```
[the remark this answers is not in the document]
Comment (Mayankbht) [reply mbed_app.json]: Done.

**Comment by Mayankbht** [comment]
run-ci

File: README.md [hunk 24]
```diff
@@ -412,3 +425,27 @@ Note platforms will be tested regularly in the Arm CI system. Please discuss wit
 # Known-issues
 
 Please review existing issues on [GitHub](https://github.com/ARMmbed/mbed-os-example-pelion/issues) and report any problem you may see.
+
+# Migration Guide for Targets supporting FOTA
```
Comment (teetak01) [inline README.md:429]:
```suggestion
# Migration Guide for Targets supporting FOTA (WIP to be replaced with official documentation).
```

File: README.md [hunk 25]
```diff
@@ -412,3 +425,27 @@ Note platforms will be tested regularly in the Arm CI system. Please discuss wit
...
+       For e.g. for K66F target following needs to be done as the FLASH size is
+       2M:
+       #truncate <PATH>/mbed-os-example-pelion.bin --size 2M
+    b. Once the BIN is truncated to the desired size, it needs to be flashed to 
+       board.
+    c. Once the BIN is flashed succefully the bootloader is upgraded to FOTA
+       supported bootloader. Now FOTA can be used to upgrade the application
+
+Note: On "How to upgrade the application using the FOTA" please refer
+    https://developer.pelion.com/docs/pelion-client-lite/v1.2/developer-guide/index.html
```
Comment (teetak01) [inline README.md:450]:
```suggestion
1. Compilation for the above FOTA supported targets is similar to as UC
   bootloader. Steps would be to:
    a. Install the mbed cli as for other targets, please check sections 
       ## Requirements
       ## Deploying
       ## Preparing for build
2. Once the above binaries are ready following needs to be done to update the
   binaries to the TARGET's
    a. The binary need to be truncated to the FLASH size as the KV store needs
       to be reset to update the KEYS on the board.
       For e.g. for K66F target following needs to be done as the FLASH size is
       2M:
       #truncate <PATH>/mbed-os-example-pelion.bin --size 2M
    b. Once the BIN is truncated to the desired size, it needs to be flashed to 
       board.
    c. Once the BIN is flashed succefully the bootloader is upgraded to FOTA
       supported bootloader. Now FOTA can be used to upgrade the application

Note: On "How to upgrade the application using the FOTA" please refer
    https://developer.pelion.com/docs/pelion-client-lite/v1.2/developer-guide/index.html
```
1. Official documentation is still work in progress.
2. Current new bootloaders are compiled from https://github.com/PelionIoT/mbed-bootloader/tree/nanoclient feature branch which provide both the application and bootloader kvstore and update configurations.
3. The provided prebuild-bl folder contain several examples for new configuration system and will be expanded in future.

**Comment by arpitarm** [comment]
run-ci

**Comment by iot-pdmc** [comment]
# [Test run: ABORTED](https://jenkins-client.isgtesting.com/job/mbed-os-example-pelion/754/)
posted-by-tag: https://jenkins-client.isgtesting.com/job/mbed-os-example-pelion/

**Comment by arpitarm** [comment]
run-ci

**Comment by iot-pdmc** [comment]
# [Test run: ABORTED](https://jenkins-client.isgtesting.com/job/mbed-os-example-pelion/755/)
posted-by-tag: https://jenkins-client.isgtesting.com/job/mbed-os-example-pelion/

**Comment by arpitarm** [comment]
run-ci

**Comment by iot-pdmc** [comment]
# [Test run: ABORTED](https://jenkins-client.isgtesting.com/job/mbed-os-example-pelion/756/)
posted-by-tag: https://jenkins-client.isgtesting.com/job/mbed-os-example-pelion/

**Comment by arpitarm** [comment]
run-ci

**Comment by arpitarm** [comment]
run-ci

**Comment by iot-pdmc** [comment]
# [Test run: ABORTED](https://jenkins-client.isgtesting.com/job/mbed-os-example-pelion/758/)
posted-by-tag: https://jenkins-client.isgtesting.com/job/mbed-os-example-pelion/

**Comment by iot-pdmc** [comment]
# [Test run: SUCCESS](https://jenkins-client.isgtesting.com/job/mbed-os-example-pelion/759/)
***
[Build artifacts](http://pdmc-ci.s3-website-eu-west-1.amazonaws.com/?prefix=jenkins-ci/mbed-os-example-pelion/759)

posted-by-tag: https://jenkins-client.isgtesting.com/job/mbed-os-example-pelion/

**Pull request closed without merging**